File: sdk/js/src/thalllend.ts
import { Connection, ReadonlyContractResult } from './connection.js';
import { Keypair } from './keypair.js';
import { PublicKey } from './publickey.js';

const PROGRAM_SYMBOL_CANDIDATES = ['LEND', 'lend', 'THALLLEND', 'thalllend'];
const MAX_U64 = (1n << 64n) - 1n;

export interface ThallLendAccountInfo {
    deposit: bigint;
    borrow: bigint;
    healthFactorBps: bigint;
}

export interface ThallLendProtocolStats {
    totalDeposits: bigint;
    totalBorrows: bigint;
    utilizationPct: bigint;
    reserves: bigint;
}

export interface ThallLendInterestRate {
    ratePerSlot: bigint;
    utilizationPct: bigint;
    totalAvailable: bigint;
}

export interface ThallLendRateModel {
    rateScale: bigint;
    slotsPerYear: bigint;
    baseRatePerSlot: bigint;
    currentRatePerSlot: bigint;
    currentAnnualBps: bigint;
    utilizationKinkPct: bigint;
    maxRatePerSlot: bigint;
}

export interface ThallLendMarketStatus {
    paused: boolean;
    licnConfigured: boolean;
    nativeLicn: boolean;
    oracleConfigPresent: boolean;
    oracleConfigValid: boolean;
    depositCap: bigint;
    reserveFactorPct: bigint;
    collateralFactorPct: bigint;
    liquidationThresholdPct: bigint;
    liquidationBonusPct: bigint;
}

export interface ThallLendStats {
    totalDeposits: number;
    totalBorrows: number;
    availableLiquidity: number;
    utilizationPct: number;
    reserves: number;
    depositCount: number;
    borrowCount: number;
    repayCount: number;
    liquidationCount: number;
    depositCap: number;
    reserveFactorPct: number;
    licnConfigured: boolean;
    nativeLicn: boolean;
    oracleConfigPresent: boolean;
    oracleConfigValid: boolean;
    paused: boolean;
}

export interface LiquidateParams {
    borrower: PublicKey | string;
    repayAmount: number | bigint;
}

function normalizeAddress(value: PublicKey | string): PublicKey {
    return value instanceof PublicKey ? value : new PublicKey(value);
}

function normalizeUnsignedU64(value: number | bigint, fieldName: string): bigint {
    const normalized = typeof value === 'bigint'
        ? value
        : Number.isSafeInteger(value) && value >= 0
            ? BigInt(value)
            : null;

    if (normalized === null || normalized < 0n || normalized > MAX_U64) {
        throw new Error(`${fieldName} must be a u64-safe integer value`);
    }

    return normalized;
}

function u64LE(value: number | bigint, fieldName: string): Uint8Array {
    const out = new Uint8Array(8);
    new DataView(out.buffer).setBigUint64(0, normalizeUnsignedU64(value, fieldName), true);
    return out;
}

function buildLayoutArgs(layout: number[], chunks: Uint8Array[]): Uint8Array {
    const header = Uint8Array.from([0xAB, ...layout]);
    const total = chunks.reduce((sum, chunk) => sum + chunk.length, header.length);
    const out = new Uint8Array(total);
    out.set(header, 0);
    let offset = header.length;
    for (const chunk of chunks) {
        out.set(chunk, offset);
        offset += chunk.length;
    }
    return out;
}

function encodeUserAmountArgs(user: PublicKey, amount: number | bigint): Uint8Array {
    return buildLayoutArgs([0x20, 0x08], [
        user.toBytes(),
        u64LE(amount, 'amount'),
    ]);
}

function encodeUserLookupArgs(user: PublicKey | string): Uint8Array {
    return buildLayoutArgs([0x20], [normalizeAddress(user).toBytes()]);
}

function encodeLiquidateArgs(liquidator: PublicKey, params: LiquidateParams): Uint8Array {
    return buildLayoutArgs([0x20, 0x20, 0x08], [
        liquidator.toBytes(),
        normalizeAddress(params.borrower).toBytes(),
        u64LE(params.repayAmount, 'repayAmount'),
    ]);
}

function decodeReturnData(returnData: string): Uint8Array {
    return Uint8Array.from(Buffer.from(returnData, 'base64'));
}

function readU64(bytes: Uint8Array, offset: number): bigint {
    const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    return view.getBigUint64(offset, true);
}

function ensureReadonlySuccess(
    result: ReadonlyContractResult,
    functionName: string,
    allowedReturnCodes: number[] = [0],
): void {
    const code = result.returnCode ?? 0;
    if (!allowedReturnCodes.includes(code)) {
        throw new Error(result.error ?? `ThallLend ${functionName} returned code ${code}`);
    }
    if (result.success === false && result.error) {
        throw new Error(result.error);
    }
}

function decodeU64Return(result: ReadonlyContractResult, functionName: string): bigint {
    ensureReadonlySuccess(result, functionName);
    if (!result.returnData) {
        throw new Error(`ThallLend ${functionName} did not return payload data`);
    }
    const bytes = decodeReturnData(result.returnData);
    if (bytes.length < 8) {
        throw new Error(`ThallLend ${functionName} payload was shorter than expected`);
    }
    return readU64(bytes, 0);
}

function decodeAccountInfo(result: ReadonlyContractResult): ThallLendAccountInfo {
    ensureReadonlySuccess(result, 'get_account_info');
    if (!result.returnData) {
        throw new Error('ThallLend get_account_info did not return account data');
    }
    const bytes = decodeReturnData(result.returnData);
    if (bytes.length < 24) {
        throw new Error('ThallLend get_account_info payload was shorter than expected');
    }
    return {
        deposit: readU64(bytes, 0),
        borrow: readU64(bytes, 8),
        healthFactorBps: readU64(bytes, 16),
    };
}

function decodeProtocolStats(result: ReadonlyContractResult): ThallLendProtocolStats {
    ensureReadonlySuccess(result, 'get_protocol_stats');
    if (!result.returnData) {
        throw new Error('ThallLend get_protocol_stats did not return stats data');
    }
    const bytes = decodeReturnData(result.returnData);
    if (bytes.length < 32) {
        throw new Error('ThallLend get_protocol_stats payload was shorter than expected');
    }
    return {
        totalDeposits: readU64(bytes, 0),
        totalBorrows: readU64(bytes, 8),
        utilizationPct: readU64(bytes, 16),
        reserves: readU64(bytes, 24),
    };
}

function decodeInterestRate(result: ReadonlyContractResult): ThallLendInterestRate {
    ensureReadonlySuccess(result, 'get_interest_rate');
    if (!result.returnData) {
        throw new Error('ThallLend get_interest_rate did not return rate data');
    }
    const bytes = decodeReturnData(result.returnData);
    if (bytes.length < 24) {
        throw new Error('ThallLend get_interest_rate payload was shorter than expected');
    }
    return {
        ratePerSlot: readU64(bytes, 0),
        utilizationPct: readU64(bytes, 8),
        totalAvailable: readU64(bytes, 16),
    };
}

function decodeRateModel(result: ReadonlyContractResult): ThallLendRateModel {
    ensureReadonlySuccess(result, 'get_rate_model');
    if (!result.returnData) {
        throw new Error('ThallLend get_rate_model did not return model data');
    }
    const bytes = decodeReturnData(result.returnData);
    if (bytes.length < 56) {
        throw new Error('ThallLend get_rate_model payload was shorter than expected');
    }
    return {
        rateScale: readU64(bytes, 0),
        slotsPerYear: readU64(bytes, 8),
        baseRatePerSlot: readU64(bytes, 16),
        currentRatePerSlot: readU64(bytes, 24),
        currentAnnualBps: readU64(bytes, 32),
        utilizationKinkPct: readU64(bytes, 40),
        maxRatePerSlot: readU64(bytes, 48),
    };
}

function decodeMarketStatus(result: ReadonlyContractResult): ThallLendMarketStatus {
    ensureReadonlySuccess(result, 'get_market_status');
    if (!result.returnData) {
        throw new Error('ThallLend get_market_status did not return status data');
    }
    const bytes = decodeReturnData(result.returnData);
    if (bytes.length < 80) {
        throw new Error('ThallLend get_market_status payload was shorter than expected');
    }
    return {
        paused: readU64(bytes, 0) !== 0n,
        licnConfigured: readU64(bytes, 8) !== 0n,
        nativeLicn: readU64(bytes, 16) !== 0n,
        oracleConfigPresent: readU64(bytes, 24) !== 0n,
        oracleConfigValid: readU64(bytes, 32) !== 0n,
        depositCap: readU64(bytes, 40),
        reserveFactorPct: readU64(bytes, 48),
        collateralFactorPct: readU64(bytes, 56),
        liquidationThresholdPct: readU64(bytes, 64),
        liquidationBonusPct: readU64(bytes, 72),
    };
}

export class ThallLendClient {
    private resolvedProgram?: PublicKey;

    constructor(
        private readonly connection: Connection,
        programId?: PublicKey,
    ) {
        this.resolvedProgram = programId;
    }

    private async callReadonly(functionName: string, args: Uint8Array = new Uint8Array()): Promise<ReadonlyContractResult> {
        const programId = await this.getProgramId();
        return this.connection.callReadonlyContract(programId, functionName, args);
    }

    async getProgramId(): Promise<PublicKey> {
        if (this.resolvedProgram) {
            return this.resolvedProgram;
        }

        for (const symbol of PROGRAM_SYMBOL_CANDIDATES) {
            try {
                const entry = await this.connection.getSymbolRegistry(symbol);
                if (entry?.program) {
                    this.resolvedProgram = new PublicKey(entry.program);
                    return this.resolvedProgram;
                }
            } catch {
                // Try the next known registry alias.
            }
        }

        throw new Error('Unable to resolve the ThallLend program via getSymbolRegistry("LEND")');
    }

    async getAccountInfo(user: PublicKey | string): Promise<ThallLendAccountInfo> {
        return decodeAccountInfo(await this.callReadonly('get_account_info', encodeUserLookupArgs(user)));
    }

    async getProtocolStats(): Promise<ThallLendProtocolStats> {
        return decodeProtocolStats(await this.callReadonly('get_protocol_stats'));
    }

    async getInterestRate(): Promise<ThallLendInterestRate> {
        return decodeInterestRate(await this.callReadonly('get_interest_rate'));
    }

    async getRateModel(): Promise<ThallLendRateModel> {
        return decodeRateModel(await this.callReadonly('get_rate_model'));
    }

    async getMarketStatus(): Promise<ThallLendMarketStatus> {
        return decodeMarketStatus(await this.callReadonly('get_market_status'));
    }

    async getDepositCount(): Promise<bigint> {
        return decodeU64Return(await this.callReadonly('get_deposit_count'), 'get_deposit_count');
    }

    async getBorrowCount(): Promise<bigint> {
        return decodeU64Return(await this.callReadonly('get_borrow_count'), 'get_borrow_count');
    }

    async getLiquidationCount(): Promise<bigint> {
        return decodeU64Return(await this.callReadonly('get_liquidation_count'), 'get_liquidation_count');
    }

    async getStats(): Promise<ThallLendStats> {
        const stats = await this.connection.getThallLendStats();
        return {
            totalDeposits: stats.total_deposits ?? 0,
            totalBorrows: stats.total_borrows ?? 0,
            availableLiquidity: stats.available_liquidity ?? 0,
            utilizationPct: stats.utilization_pct ?? 0,
            reserves: stats.reserves ?? 0,
            depositCount: stats.deposit_count ?? 0,
            borrowCount: stats.borrow_count ?? 0,
            repayCount: stats.repay_count ?? 0,
            liquidationCount: stats.liquidation_count ?? 0,
            depositCap: stats.deposit_cap ?? 0,
            reserveFactorPct: stats.reserve_factor_pct ?? 0,
            licnConfigured: Boolean(stats.licn_configured),
            nativeLicn: Boolean(stats.native_licn),
            oracleConfigPresent: Boolean(stats.oracle_config_present),
            oracleConfigValid: Boolean(stats.oracle_config_valid),
            paused: Boolean(stats.paused),
        };
    }

    async deposit(depositor: Keypair, amount: number | bigint): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeUserAmountArgs(depositor.pubkey(), amount);
        return this.connection.callContract(depositor, programId, 'deposit', args, normalizeUnsignedU64(amount, 'amount'));
    }

    async withdraw(depositor: Keypair, amount: number | bigint): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeUserAmountArgs(depositor.pubkey(), amount);
        return this.connection.callContract(depositor, programId, 'withdraw', args);
    }

    async borrow(borrower: Keypair, amount: number | bigint): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeUserAmountArgs(borrower.pubkey(), amount);
        return this.connection.callContract(borrower, programId, 'borrow', args);
    }

    async repay(borrower: Keypair, amount: number | bigint): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeUserAmountArgs(borrower.pubkey(), amount);
        return this.connection.callContract(borrower, programId, 'repay', args, normalizeUnsignedU64(amount, 'amount'));
    }

    async liquidate(liquidator: Keypair, params: LiquidateParams): Promise<string> {
        const programId = await this.getProgramId();
        const args = encodeLiquidateArgs(liquidator.pubkey(), params);
        return this.connection.callContract(
            liquidator,
            programId,
            'liquidate',
            args,
            normalizeUnsignedU64(params.repayAmount, 'repayAmount'),
        );
    }
}

File: sdk/rust/src/thalllend.rs
use crate::client::ReadonlyContractResult;
use crate::{Client, Error, Keypair, Pubkey, Result};
use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};

const PROGRAM_SYMBOL_CANDIDATES: [&str; 4] = ["LEND", "lend", "THALLLEND", "thalllend"];
const ACCOUNT_INFO_SIZE: usize = 24;
const PROTOCOL_STATS_SIZE: usize = 32;
const INTEREST_RATE_SIZE: usize = 24;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ThallLendAccountInfo {
    pub deposit: u64,
    pub borrow: u64,
    pub health_factor_bps: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ThallLendProtocolStats {
    pub total_deposits: u64,
    pub total_borrows: u64,
    pub utilization_pct: u64,
    pub reserves: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ThallLendInterestRate {
    pub rate_per_slot: u64,
    pub utilization_pct: u64,
    pub total_available: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ThallLendStats {
    pub total_deposits: u64,
    pub total_borrows: u64,
    pub reserves: u64,
    pub deposit_count: u64,
    pub borrow_count: u64,
    pub liquidation_count: u64,
    pub paused: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LiquidateParams {
    pub borrower: Pubkey,
    pub repay_amount: u64,
}

#[derive(Debug, Clone)]
pub struct ThallLendClient {
    client: Client,
    program_id: Arc<Mutex<Option<Pubkey>>>,
}

fn build_layout_args(layout: &[u8], chunks: &[Vec<u8>]) -> Vec<u8> {
    let mut out = Vec::with_capacity(
        1 + layout.len() + chunks.iter().map(|chunk| chunk.len()).sum::<usize>(),
    );
    out.push(0xAB);
    out.extend_from_slice(layout);
    for chunk in chunks {
        out.extend_from_slice(chunk);
    }
    out
}

fn encode_user_amount_args(user: &Pubkey, amount: u64) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x08],
        &[user.as_ref().to_vec(), amount.to_le_bytes().to_vec()],
    )
}

fn encode_user_lookup_args(user: &Pubkey) -> Vec<u8> {
    build_layout_args(&[0x20], &[user.as_ref().to_vec()])
}

fn encode_liquidate_args(liquidator: &Pubkey, params: &LiquidateParams) -> Vec<u8> {
    build_layout_args(
        &[0x20, 0x20, 0x08],
        &[
            liquidator.as_ref().to_vec(),
            params.borrower.as_ref().to_vec(),
            params.repay_amount.to_le_bytes().to_vec(),
        ],
    )
}

fn ensure_readonly_success(
    result: &ReadonlyContractResult,
    function_name: &str,
    allowed_codes: &[u32],
) -> Result<()> {
    let code = result.return_code.unwrap_or(0);
    if !allowed_codes.contains(&code) {
        return Err(Error::RpcError(result.error.clone().unwrap_or_else(|| {
            format!("ThallLend {} returned code {}", function_name, code)
        })));
    }
    if !result.success {
        return Err(Error::RpcError(
            result
                .error
                .clone()
                .unwrap_or_else(|| format!("ThallLend {} failed", function_name)),
        ));
    }
    Ok(())
}

fn decode_return_data(result: &ReadonlyContractResult, function_name: &str) -> Result<Vec<u8>> {
    let Some(return_data) = &result.return_data else {
        return Err(Error::ParseError(format!(
            "ThallLend {} did not return payload data",
            function_name,
        )));
    };

    base64::Engine::decode(&base64::engine::general_purpose::STANDARD, return_data)
        .map_err(|err| Error::ParseError(err.to_string()))
}

fn decode_u64(bytes: &[u8], start: usize, function_name: &str) -> Result<u64> {
    let end = start + 8;
    if bytes.len() < end {
        return Err(Error::ParseError(format!(
            "ThallLend {} payload was shorter than expected",
            function_name,
        )));
    }
    let slice: [u8; 8] = bytes[start..end].try_into().map_err(|_| {
        Error::ParseError(format!("ThallLend {} payload was malformed", function_name))
    })?;
    Ok(u64::from_le_bytes(slice))
}

fn decode_u64_result(result: &ReadonlyContractResult, function_name: &str) -> Result<u64> {
    ensure_readonly_success(result, function_name, &[0])?;
    let bytes = decode_return_data(result, function_name)?;
    decode_u64(&bytes, 0, function_name)
}

fn decode_account_info(result: &ReadonlyContractResult) -> Result<ThallLendAccountInfo> {
    ensure_readonly_success(result, "get_account_info", &[0])?;
    let bytes = decode_return_data(result, "get_account_info")?;
    if bytes.len() < ACCOUNT_INFO_SIZE {
        return Err(Error::ParseError(
            "ThallLend get_account_info payload was shorter than expected".into(),
        ));
    }

    Ok(ThallLendAccountInfo {
        deposit: decode_u64(&bytes, 0, "get_account_info")?,
        borrow: decode_u64(&bytes, 8, "get_account_info")?,
        health_factor_bps: decode_u64(&bytes, 16, "get_account_info")?,
    })
}

fn decode_protocol_stats(result: &ReadonlyContractResult) -> Result<ThallLendProtocolStats> {
    ensure_readonly_success(result, "get_protocol_stats", &[0])?;
    let bytes = decode_return_data(result, "get_protocol_stats")?;
    if bytes.len() < PROTOCOL_STATS_SIZE {
        return Err(Error::ParseError(
            "ThallLend get_protocol_stats payload was shorter than expected".into(),
        ));
    }

    Ok(ThallLendProtocolStats {
        total_deposits: decode_u64(&bytes, 0, "get_protocol_stats")?,
        total_borrows: decode_u64(&bytes, 8, "get_protocol_stats")?,
        utilization_pct: decode_u64(&bytes, 16, "get_protocol_stats")?,
        reserves: decode_u64(&bytes, 24, "get_protocol_stats")?,
    })
}

fn decode_interest_rate(result: &ReadonlyContractResult) -> Result<ThallLendInterestRate> {
    ensure_readonly_success(result, "get_interest_rate", &[0])?;
    let bytes = decode_return_data(result, "get_interest_rate")?;
    if bytes.len() < INTEREST_RATE_SIZE {
        return Err(Error::ParseError(
            "ThallLend get_interest_rate payload was shorter than expected".into(),
        ));
    }

    Ok(ThallLendInterestRate {
        rate_per_slot: decode_u64(&bytes, 0, "get_interest_rate")?,
        utilization_pct: decode_u64(&bytes, 8, "get_interest_rate")?,
        total_available: decode_u64(&bytes, 16, "get_interest_rate")?,
    })
}

impl ThallLendClient {
    pub fn new(client: Client) -> Self {
        Self {
            client,
            program_id: Arc::new(Mutex::new(None)),
        }
    }

    pub fn with_program_id(client: Client, program_id: Pubkey) -> Self {
        Self {
            client,
            program_id: Arc::new(Mutex::new(Some(program_id))),
        }
    }

    pub async fn get_program_id(&self) -> Result<Pubkey> {
        if let Some(program_id) = self
            .program_id
            .lock()
            .map_err(|_| Error::ConfigError("ThallLendClient program cache lock poisoned".into()))?
            .clone()
        {
            return Ok(program_id);
        }

        for symbol in PROGRAM_SYMBOL_CANDIDATES {
            let entry = match self.client.get_symbol_registry(symbol).await {
                Ok(entry) => entry,
                Err(_) => continue,
            };
            let Some(program) = entry.get("program").and_then(|value| value.as_str()) else {
                continue;
            };
            let program_id = Pubkey::from_base58(program).map_err(Error::ParseError)?;
            *self.program_id.lock().map_err(|_| {
                Error::ConfigError("ThallLendClient program cache lock poisoned".into())
            })? = Some(program_id);
            return Ok(program_id);
        }

        Err(Error::ConfigError(
            "Unable to resolve the ThallLend program via getSymbolRegistry(\"LEND\")".into(),
        ))
    }

    pub async fn get_account_info(&self, user: &Pubkey) -> Result<ThallLendAccountInfo> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_account_info",
                encode_user_lookup_args(user),
                None,
            )
            .await?;
        decode_account_info(&result)
    }

    pub async fn get_protocol_stats(&self) -> Result<ThallLendProtocolStats> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_protocol_stats",
                Vec::new(),
                None,
            )
            .await?;
        decode_protocol_stats(&result)
    }

    pub async fn get_interest_rate(&self) -> Result<ThallLendInterestRate> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_interest_rate",
                Vec::new(),
                None,
            )
            .await?;
        decode_interest_rate(&result)
    }

    pub async fn get_deposit_count(&self) -> Result<u64> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_deposit_count",
                Vec::new(),
                None,
            )
            .await?;
        decode_u64_result(&result, "get_deposit_count")
    }

    pub async fn get_borrow_count(&self) -> Result<u64> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_borrow_count",
                Vec::new(),
                None,
            )
            .await?;
        decode_u64_result(&result, "get_borrow_count")
    }

    pub async fn get_liquidation_count(&self) -> Result<u64> {
        let result = self
            .client
            .call_readonly_contract(
                &self.get_program_id().await?,
                "get_liquidation_count",
                Vec::new(),
                None,
            )
            .await?;
        decode_u64_result(&result, "get_liquidation_count")
    }

    pub async fn get_stats(&self) -> Result<ThallLendStats> {
        let value = self.client.get_thalllend_stats().await?;
        serde_json::from_value(value).map_err(|err| Error::ParseError(err.to_string()))
    }

    pub async fn deposit(&self, depositor: &Keypair, amount: u64) -> Result<String> {
        let program_id = self.get_program_id().await?;
        self.client
            .call_contract(
                depositor,
                &program_id,
                "deposit",
                encode_user_amount_args(&depositor.pubkey(), amount),
                amount,
            )
            .await
    }

    pub async fn withdraw(&self, depositor: &Keypair, amount: u64) -> Result<String> {
        let program_id = self.get_program_id().await?;
        self.client
            .call_contract(
                depositor,
                &program_id,
                "withdraw",
                encode_user_amount_args(&depositor.pubkey(), amount),
                0,
            )
            .await
    }

    pub async fn borrow(&self, borrower: &Keypair, amount: u64) -> Result<String> {
        let program_id = self.get_program_id().await?;
        self.client
            .call_contract(
                borrower,
                &program_id,
                "borrow",
                encode_user_amount_args(&borrower.pubkey(), amount),
                0,
            )
            .await
    }

    pub async fn repay(&self, borrower: &Keypair, amount: u64) -> Result<String> {
        let program_id = self.get_program_id().await?;
        self.client
            .call_contract(
                borrower,
                &program_id,
                "repay",
                encode_user_amount_args(&borrower.pubkey(), amount),
                amount,
            )
            .await
    }

    pub async fn liquidate(&self, liquidator: &Keypair, params: LiquidateParams) -> Result<String> {
        let program_id = self.get_program_id().await?;
        self.client
            .call_contract(
                liquidator,
                &program_id,
                "liquidate",
                encode_liquidate_args(&liquidator.pubkey(), &params),
                params.repay_amount,
            )
            .await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn readonly_result(bytes: Vec<u8>) -> ReadonlyContractResult {
        ReadonlyContractResult {
            success: true,
            return_data: Some(base64::Engine::encode(
                &base64::engine::general_purpose::STANDARD,
                bytes,
            )),
            return_code: Some(0),
            logs: Vec::new(),
            error: None,
            compute_used: None,
        }
    }

    #[test]
    fn user_amount_encoding_matches_named_export_layout() {
        let user = Pubkey([7u8; 32]);
        let encoded = encode_user_amount_args(&user, 1_000);

        assert_eq!(&encoded[..3], &[0xAB, 0x20, 0x08]);
        assert_eq!(&encoded[3..35], &[7u8; 32]);
        assert_eq!(
            u64::from_le_bytes(encoded[35..43].try_into().unwrap()),
            1_000
        );
    }

    #[test]
    fn liquidate_encoding_includes_borrower_and_repay_amount() {
        let liquidator = Pubkey([8u8; 32]);
        let params = LiquidateParams {
            borrower: Pubkey([9u8; 32]),
            repay_amount: 250,
        };

        let encoded = encode_liquidate_args(&liquidator, &params);

        assert_eq!(&encoded[..4], &[0xAB, 0x20, 0x20, 0x08]);
        assert_eq!(&encoded[4..36], &[8u8; 32]);
        assert_eq!(&encoded[36..68], &[9u8; 32]);
        assert_eq!(u64::from_le_bytes(encoded[68..76].try_into().unwrap()), 250);
    }

    #[test]
    fn account_info_decoding_matches_contract_layout() {
        let mut payload = Vec::new();
        payload.extend_from_slice(&1_000_u64.to_le_bytes());
        payload.extend_from_slice(&400_u64.to_le_bytes());
        payload.extend_from_slice(&21_250_u64.to_le_bytes());

        let decoded = decode_account_info(&readonly_result(payload)).unwrap();

        assert_eq!(decoded.deposit, 1_000);
        assert_eq!(decoded.borrow, 400);
        assert_eq!(decoded.health_factor_bps, 21_250);
    }

    #[test]
    fn protocol_stats_and_interest_rate_decoding_match_contract_layouts() {
        let mut protocol_payload = Vec::new();
        protocol_payload.extend_from_slice(&5_000_u64.to_le_bytes());
        protocol_payload.extend_from_slice(&2_000_u64.to_le_bytes());
        protocol_payload.extend_from_slice(&40_u64.to_le_bytes());
        protocol_payload.extend_from_slice(&150_u64.to_le_bytes());

        let protocol = decode_protocol_stats(&readonly_result(protocol_payload)).unwrap();
        assert_eq!(protocol.total_deposits, 5_000);
        assert_eq!(protocol.total_borrows, 2_000);
        assert_eq!(protocol.utilization_pct, 40);
        assert_eq!(protocol.reserves, 150);

        let mut rate_payload = Vec::new();
        rate_payload.extend_from_slice(&254_u64.to_le_bytes());
        rate_payload.extend_from_slice(&40_u64.to_le_bytes());
        rate_payload.extend_from_slice(&3_000_u64.to_le_bytes());

        let rate = decode_interest_rate(&readonly_result(rate_payload)).unwrap();
        assert_eq!(rate.rate_per_slot, 254);
        assert_eq!(rate.utilization_pct, 40);
        assert_eq!(rate.total_available, 3_000);
    }
}
